{
 "cells": [
  {
   "cell_type": "markdown",
   "id": "213248b9",
   "metadata": {},
   "source": [
    "### Workflow\n",
    "\n",
    "Figma diagram: https://www.figma.com/file/PFKdJO3HTIcM9g4nmVONCT/HRA-Data-Explorer?type=whiteboard&node-id=1%3A108&t=XGoN91s7IKbyZ7xk-1\n",
    "\n",
    "Below is the information about the diagram:\n",
    "\n",
    "1. Dataset Collections:\n",
    "The process begins with RUI Registered Tissue datasets, a collection of tissue datasets registered in a specific repository or database.\n",
    "These datasets are linked to specific Organs and Anatomical Structures.\n",
    "\n",
    "2. Dataset Processing:\n",
    "The datasets are processed to produce Anatomical Structures/Cell type information of 2D FTUs.\n",
    "There is a decision point or a step to possibly sort these structures, which leads to the extraction of a Cell Summary, including cell type biomarker gene expression data.\n",
    "    \n",
    "    1. Compare FTU's cell type information between Azimuth, PopV and Celltypist\n",
    "    2. Get the list of dataset for FTU.\n",
    "        1. Cell summary from the atlas-as-cell-summaries.jsonld\n",
    "        2. add organ data from the cell annotation crosswalk files\n",
    "    3. Any dataset that is been shared among two FTUs\n",
    "\n",
    "3. Cell Type Mapping:\n",
    "list of 2D FTUs are used to map Cell Types, leading to a set of inquiries on how to best extract data from the datasets used in Step 1 and 2.\n",
    "\n",
    "4. Additonal work"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "c0c450b3",
   "metadata": {},
   "source": [
    "### Get the Cell annotation tool information for cell types in FTUs"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "6c48006a",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "File saved to FTU_Cell_Count_Annotated.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Supriya\\AppData\\Local\\Temp\\ipykernel_11880\\965254153.py:54: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  ftu_cell_count_filtered.loc[:, 'Organ_ID'] = ftu_cell_count_filtered['Organ'].map(organ_to_uberon)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import requests\n",
    "\n",
    "# Function to download CSV files from GitHub\n",
    "def download_csv_from_github(url, local_path):\n",
    "    response = requests.get(url)\n",
    "    response.raise_for_status()  # Check if the request was successful\n",
    "    with open(local_path, 'wb') as file:\n",
    "        file.write(response.content)\n",
    "\n",
    "# URLs of the CSV files in the GitHub repository\n",
    "# github_base_url = 'https://github.com/hubmapconsortium/hra-workflows-runner/blob/main/crosswalking-tables/'\n",
    "# popv_url = github_base_url + 'popv.csv'\n",
    "# celltypist_url = github_base_url + 'celltypist.csv'\n",
    "# azimuth_url = github_base_url + 'azimuth.csv'\n",
    "# ftu_cell_count_url = github_base_url + 'FTU_Cell_Count_Table_Cell_Type_Count.csv'\n",
    "\n",
    "# Local paths to save the downloaded files\n",
    "popv_local_path = 'C:\\\\Users\\\\Supriya\\\\Downloads\\\\popv.csv'\n",
    "celltypist_local_path = 'C:\\\\Users\\\\Supriya\\\\Downloads\\\\celltypist.csv'\n",
    "azimuth_local_path = 'C:\\\\Users\\\\Supriya\\\\Downloads\\\\azimuth.csv'\n",
    "ftu_cell_count_path = 'C:\\\\Users\\\\Supriya\\\\Downloads\\\\FTU Cell Count Table - Cell_Type_Count.csv'\n",
    "\n",
    "\n",
    "# Download the files\n",
    "#download_csv_from_github(popv_url, popv_local_path)\n",
    "#download_csv_from_github(celltypist_url, celltypist_local_path)\n",
    "#download_csv_from_github(azimuth_url, azimuth_local_path)\n",
    "\n",
    "# Load the downloaded files\n",
    "popv = pd.read_csv(popv_local_path)\n",
    "celltypist = pd.read_csv(celltypist_local_path)\n",
    "azimuth = pd.read_csv(azimuth_local_path)\n",
    "ftu_cell_count = pd.read_csv(ftu_cell_count_path)\n",
    "\n",
    "# Define the mapping dictionary\n",
    "organ_to_uberon = {\n",
    "    'Kidney': 'UBERON:0002113',\n",
    "    'Lung': 'UBERON:0002048',\n",
    "    'Pancreas': 'UBERON:0001264',\n",
    "    'Large Intestine': 'UBERON:0002107',\n",
    "    'Skin': 'UBERON:0002097',\n",
    "    'Liver': 'UBERON:0002108',\n",
    "    'Prostate': 'UBERON:0002367',\n",
    "    'Thymus': 'UBERON:0002371',\n",
    "    'Spleen': 'UBERON:0002370',\n",
    "    'Small Intestine': 'UBERON:0002106'\n",
    "}\n",
    "\n",
    "# Filter FTU cell count to include only rows where `CT ID in CL` contains `CL`\n",
    "ftu_cell_count_filtered = ftu_cell_count[ftu_cell_count['CT ID in CL'].str.contains('CL', na=False)]\n",
    "\n",
    "# Map the organ names in FTU cell count table to UBERON IDs\n",
    "ftu_cell_count_filtered.loc[:, 'Organ_ID'] = ftu_cell_count_filtered['Organ'].map(organ_to_uberon)\n",
    "\n",
    "# Prepare a common structure to hold the results\n",
    "result = []\n",
    "\n",
    "# Function to dynamically generate match info\n",
    "def generate_match_info(tool_name, tool_data):\n",
    "    levels = set()\n",
    "    for organ_level in tool_data['Organ_Level']:\n",
    "        level = organ_level.split('_')[-1]\n",
    "        levels.add(level)\n",
    "    \n",
    "    match_info = {f'{tool_name}_{level}': 0 for level in levels}\n",
    "    return match_info, levels\n",
    "\n",
    "# Dynamically generate the match info for each tool\n",
    "celltypist_info, celltypist_levels = generate_match_info('celltypist', celltypist)\n",
    "azimuth_info, azimuth_levels = generate_match_info('azimuth', azimuth)\n",
    "popv_info, popv_levels = generate_match_info('popv', popv)\n",
    "\n",
    "# Iterate over the filtered FTU cell count data\n",
    "for idx, ftu_row in ftu_cell_count_filtered.iterrows():\n",
    "    organ = ftu_row['Organ_ID']\n",
    "    ftu_label = ftu_row['FTU Label in Uberon']\n",
    "    ftu_id = ftu_row['FTU ID in Uberon']\n",
    "    cl_id = ftu_row['CT ID in CL']\n",
    "    cl_label = ftu_row['CT Label in CL']\n",
    "    \n",
    "    # Initialize match columns\n",
    "    match_info = {\n",
    "        'Organ': ftu_row['Organ'],\n",
    "        'FTU Label in Uberon': ftu_label,\n",
    "        'FTU ID in Uberon': ftu_id,\n",
    "        'CL_id': cl_id,\n",
    "        'CT Label in CL': cl_label,\n",
    "    }\n",
    "    match_info.update({key: 0 for key in celltypist_info.keys()})\n",
    "    match_info.update({key: 0 for key in azimuth_info.keys()})\n",
    "    match_info.update({key: 0 for key in popv_info.keys()})\n",
    "    \n",
    "    # Check in celltypist\n",
    "    celltypist_matches = celltypist[(celltypist['CL_ID'] == cl_id) & (celltypist['Organ_ID'] == organ)]\n",
    "    for _, ct_match in celltypist_matches.iterrows():\n",
    "        level = ct_match['Organ_Level'].split('_')[-1]  # Extract the level number\n",
    "        match_info[f'celltypist_{level}'] = 1\n",
    "\n",
    "    # Check in azimuth\n",
    "    azimuth_matches = azimuth[(azimuth['CL_ID'] == cl_id) & (azimuth['Organ_ID'] == organ)]\n",
    "    for _, az_match in azimuth_matches.iterrows():\n",
    "        level = az_match['Organ_Level'].split('_')[-1]  # Extract the level number\n",
    "        match_info[f'azimuth_{level}'] = 1\n",
    "\n",
    "    # Check in popv\n",
    "    popv_matches = popv[(popv['CL_ID'] == cl_id) & (popv['Organ_ID'] == organ)]\n",
    "    if not popv_matches.empty:\n",
    "        for level in popv_levels:\n",
    "            match_info[f'popv_{level}'] = 1\n",
    "\n",
    "    # Append to results\n",
    "    result.append(match_info)\n",
    "\n",
    "# Create a dataframe from the results\n",
    "result_df = pd.DataFrame(result)\n",
    "\n",
    "# Save the result to a CSV file\n",
    "output_path = 'FTU_Cell_Count_Annotated.csv'\n",
    "result_df.to_csv(output_path, index=False)\n",
    "\n",
    "print(f\"File saved to {output_path}\")\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 6,
   "id": "35dec1cb",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Load all the files into dataframes\n",
    "atlas_as_cell_summaries_path = 'ref_data/atlas-as-cell-summaries.jsonld'\n",
    "\n",
    "with open(atlas_as_cell_summaries_path, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "# Function to extract structure and keys of the JSON\n",
    "def extract_structure(data, level=0):\n",
    "    structure = {}\n",
    "    if isinstance(data, dict):\n",
    "        for key, value in data.items():\n",
    "            structure[key] = extract_structure(value, level + 1)\n",
    "    elif isinstance(data, list) and len(data) > 0:\n",
    "        structure = [extract_structure(data[0], level + 1)]\n",
    "    else:\n",
    "        structure = None\n",
    "    return structure\n",
    "\n",
    "# Extract structure\n",
    "json_structure_cell_summaries = extract_structure(json_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 7,
   "id": "c2fac17f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'@context': {'CL': {'@id': None, '@prefix': None},\n",
       "  'ASCTB-TEMP': {'@id': None, '@prefix': None},\n",
       "  'ctpop': {'@id': None, '@prefix': None},\n",
       "  'as_3d_id': {'@type': None},\n",
       "  'as_id': {'@type': None},\n",
       "  'all_collisions': {'@id': None},\n",
       "  'collision_source': {'@reverse': None, '@type': None},\n",
       "  'collisions': {'@id': None},\n",
       "  'corridor_source': {'@reverse': None, '@type': None},\n",
       "  'corridor': {'@id': None},\n",
       "  'summaries': {'@id': None},\n",
       "  'cell_source': {'@reverse': None, '@type': None},\n",
       "  'aggregated_summaries': {'@id': None, '@type': None},\n",
       "  'annotation_method': {'@id': None},\n",
       "  'summary': {'@id': None},\n",
       "  'cell_id': {'@type': None},\n",
       "  'count': {'@id': None},\n",
       "  'percentage': {'@id': None},\n",
       "  'cell_count': {'@id': None, '@type': None},\n",
       "  'gene_count': {'@id': None, '@type': None},\n",
       "  'organ_id': {'@type': None},\n",
       "  'cell_source_a': {'@type': None},\n",
       "  'cell_source_b': {'@type': None},\n",
       "  'entity_a': {'@type': None},\n",
       "  'entity_b': {'@type': None},\n",
       "  '@base': None,\n",
       "  '@vocab': None,\n",
       "  'ccf': None,\n",
       "  'rdfs': None,\n",
       "  'dcterms': None,\n",
       "  'label': None,\n",
       "  'description': None,\n",
       "  'link': None,\n",
       "  'sex': None,\n",
       "  'age': None,\n",
       "  'bmi': None,\n",
       "  'consortium_name': None,\n",
       "  'provider_name': None,\n",
       "  'provider_uuid': None,\n",
       "  'donor': {'@id': None, '@type': None},\n",
       "  'samples': {'@reverse': None},\n",
       "  'sections': {'@id': None, '@type': None},\n",
       "  'datasets': {'@id': None, '@type': None},\n",
       "  'sample_type': None,\n",
       "  'section_count': None,\n",
       "  'section_size': None,\n",
       "  'section_units': None,\n",
       "  'section_number': None,\n",
       "  'rui_location': {'@id': None, '@type': None},\n",
       "  'ccf_annotations': {'@id': None, '@type': None, '@container': None},\n",
       "  'representation_of': {'@id': None, '@type': None},\n",
       "  'reference_organ': {'@id': None, '@type': None},\n",
       "  'extraction_set_for': {'@id': None, '@type': None},\n",
       "  'extraction_set': {'@id': None, '@type': None},\n",
       "  'organ_owner_sex': None,\n",
       "  'side': None,\n",
       "  'rui_rank': None,\n",
       "  'slice_thickness': None,\n",
       "  'slice_count': None,\n",
       "  'object': {'@id': None, '@type': None},\n",
       "  'creation_date': None,\n",
       "  'updated_date': None,\n",
       "  'creator': None,\n",
       "  'creator_first_name': None,\n",
       "  'creator_last_name': None,\n",
       "  'placement': {'@reverse': None},\n",
       "  'placement_date': None,\n",
       "  'rotation_order': None,\n",
       "  'dimension_units': None,\n",
       "  'rotation_units': None,\n",
       "  'scaling_units': None,\n",
       "  'translation_units': None,\n",
       "  'source': {'@id': None, '@type': None},\n",
       "  'target': {'@id': None, '@type': None},\n",
       "  'x_rotation': None,\n",
       "  'y_rotation': None,\n",
       "  'z_rotation': None,\n",
       "  'x_scaling': None,\n",
       "  'y_scaling': None,\n",
       "  'z_scaling': None,\n",
       "  'x_translation': None,\n",
       "  'y_translation': None,\n",
       "  'z_translation': None,\n",
       "  'x_dimension': None,\n",
       "  'y_dimension': None,\n",
       "  'z_dimension': None,\n",
       "  'ontology_terms': {'@id': None, '@type': None},\n",
       "  'technology': None,\n",
       "  'thumbnail': None,\n",
       "  'file': None,\n",
       "  'file_format': None,\n",
       "  'file_subpath': None},\n",
       " '@graph': [{'@type': None,\n",
       "   'sex': None,\n",
       "   'cell_source': None,\n",
       "   'cell_source_label': None,\n",
       "   'annotation_method': None,\n",
       "   'aggregated_summary_count': None,\n",
       "   'aggregated_summaries': [None],\n",
       "   'modality': None,\n",
       "   'summary': [{'@type': None,\n",
       "     'cell_id': None,\n",
       "     'cell_label': None,\n",
       "     'count': None,\n",
       "     'percentage': None}]}]}"
      ]
     },
     "execution_count": 7,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_structure_cell_summaries"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 8,
   "id": "6a1d76ec",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Load the JSON file\n",
    "json_file_path = 'ref_data/atlas-enriched-dataset-graph.jsonld'\n",
    "with open(json_file_path, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "# Function to extract structure and keys of the JSON\n",
    "def extract_structure(data, level=0):\n",
    "    structure = {}\n",
    "    if isinstance(data, dict):\n",
    "        for key, value in data.items():\n",
    "            structure[key] = extract_structure(value, level + 1)\n",
    "    elif isinstance(data, list) and len(data) > 0:\n",
    "        structure = [extract_structure(data[0], level + 1)]\n",
    "    else:\n",
    "        structure = None\n",
    "    return structure\n",
    "\n",
    "# Extract structure\n",
    "json_structure = extract_structure(json_data)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 9,
   "id": "ea5f2c59",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'@context': {'CL': {'@id': None, '@prefix': None},\n",
       "  'ASCTB-TEMP': {'@id': None, '@prefix': None},\n",
       "  'ctpop': {'@id': None, '@prefix': None},\n",
       "  'as_3d_id': {'@type': None},\n",
       "  'as_id': {'@type': None},\n",
       "  'all_collisions': {'@id': None},\n",
       "  'collision_source': {'@reverse': None, '@type': None},\n",
       "  'collisions': {'@id': None},\n",
       "  'corridor_source': {'@reverse': None, '@type': None},\n",
       "  'corridor': {'@id': None},\n",
       "  'summaries': {'@id': None},\n",
       "  'cell_source': {'@reverse': None, '@type': None},\n",
       "  'aggregated_summaries': {'@id': None, '@type': None},\n",
       "  'annotation_method': {'@id': None},\n",
       "  'summary': {'@id': None},\n",
       "  'cell_id': {'@type': None},\n",
       "  'count': {'@id': None},\n",
       "  'percentage': {'@id': None},\n",
       "  'cell_count': {'@id': None, '@type': None},\n",
       "  'gene_count': {'@id': None, '@type': None},\n",
       "  'organ_id': {'@type': None},\n",
       "  'cell_source_a': {'@type': None},\n",
       "  'cell_source_b': {'@type': None},\n",
       "  'entity_a': {'@type': None},\n",
       "  'entity_b': {'@type': None},\n",
       "  '@base': None,\n",
       "  '@vocab': None,\n",
       "  'ccf': None,\n",
       "  'rdfs': None,\n",
       "  'dcterms': None,\n",
       "  'label': None,\n",
       "  'description': None,\n",
       "  'link': None,\n",
       "  'sex': None,\n",
       "  'age': None,\n",
       "  'bmi': None,\n",
       "  'consortium_name': None,\n",
       "  'provider_name': None,\n",
       "  'provider_uuid': None,\n",
       "  'donor': {'@id': None, '@type': None},\n",
       "  'samples': {'@reverse': None},\n",
       "  'sections': {'@id': None, '@type': None},\n",
       "  'datasets': {'@id': None, '@type': None},\n",
       "  'sample_type': None,\n",
       "  'section_count': None,\n",
       "  'section_size': None,\n",
       "  'section_units': None,\n",
       "  'section_number': None,\n",
       "  'rui_location': {'@id': None, '@type': None},\n",
       "  'ccf_annotations': {'@id': None, '@type': None, '@container': None},\n",
       "  'representation_of': {'@id': None, '@type': None},\n",
       "  'reference_organ': {'@id': None, '@type': None},\n",
       "  'extraction_set_for': {'@id': None, '@type': None},\n",
       "  'extraction_set': {'@id': None, '@type': None},\n",
       "  'organ_owner_sex': None,\n",
       "  'side': None,\n",
       "  'rui_rank': None,\n",
       "  'slice_thickness': None,\n",
       "  'slice_count': None,\n",
       "  'object': {'@id': None, '@type': None},\n",
       "  'creation_date': None,\n",
       "  'updated_date': None,\n",
       "  'creator': None,\n",
       "  'creator_first_name': None,\n",
       "  'creator_last_name': None,\n",
       "  'placement': {'@reverse': None},\n",
       "  'placement_date': None,\n",
       "  'rotation_order': None,\n",
       "  'dimension_units': None,\n",
       "  'rotation_units': None,\n",
       "  'scaling_units': None,\n",
       "  'translation_units': None,\n",
       "  'source': {'@id': None, '@type': None},\n",
       "  'target': {'@id': None, '@type': None},\n",
       "  'x_rotation': None,\n",
       "  'y_rotation': None,\n",
       "  'z_rotation': None,\n",
       "  'x_scaling': None,\n",
       "  'y_scaling': None,\n",
       "  'z_scaling': None,\n",
       "  'x_translation': None,\n",
       "  'y_translation': None,\n",
       "  'z_translation': None,\n",
       "  'x_dimension': None,\n",
       "  'y_dimension': None,\n",
       "  'z_dimension': None,\n",
       "  'ontology_terms': {'@id': None, '@type': None},\n",
       "  'technology': None,\n",
       "  'thumbnail': None,\n",
       "  'file': None,\n",
       "  'file_format': None,\n",
       "  'file_subpath': None},\n",
       " '@graph': [{'@id': None,\n",
       "   '@type': None,\n",
       "   'samples': [{'@id': None,\n",
       "     '@type': None,\n",
       "     'datasets': None,\n",
       "     'sections': [{'@id': None,\n",
       "       '@type': None,\n",
       "       'datasets': [{'@id': None,\n",
       "         '@type': None,\n",
       "         'link': None,\n",
       "         'technology': None,\n",
       "         'cell_count': None,\n",
       "         'gene_count': None,\n",
       "         'organ_id': None,\n",
       "         'label': None,\n",
       "         'description': None,\n",
       "         'thumbnail': None,\n",
       "         'summaries': [{'@type': None,\n",
       "           'annotation_method': None,\n",
       "           'modality': None,\n",
       "           'summary': [{'cell_id': None,\n",
       "             'cell_label': None,\n",
       "             'gene_expr': [{'gene_id': None,\n",
       "               'gene_label': None,\n",
       "               'ensembl_id': None,\n",
       "               'mean_gene_expr_value': None}],\n",
       "             'count': None,\n",
       "             '@type': None,\n",
       "             'percentage': None}],\n",
       "           'sex': None}]}],\n",
       "       'sample_type': None,\n",
       "       'link': None,\n",
       "       'label': None,\n",
       "       'description': None}],\n",
       "     'sample_type': None,\n",
       "     'link': None,\n",
       "     'label': None,\n",
       "     'description': None,\n",
       "     'rui_location': {'@context': None,\n",
       "      '@id': None,\n",
       "      '@type': None,\n",
       "      'ccf_annotations': [None],\n",
       "      'creation_date': None,\n",
       "      'creator': None,\n",
       "      'creator_first_name': None,\n",
       "      'creator_last_name': None,\n",
       "      'dimension_units': None,\n",
       "      'placement': {'@context': None,\n",
       "       '@id': None,\n",
       "       '@type': None,\n",
       "       'placement_date': None,\n",
       "       'rotation_order': None,\n",
       "       'rotation_units': None,\n",
       "       'scaling_units': None,\n",
       "       'target': None,\n",
       "       'translation_units': None,\n",
       "       'x_rotation': None,\n",
       "       'x_scaling': None,\n",
       "       'x_translation': None,\n",
       "       'y_rotation': None,\n",
       "       'y_scaling': None,\n",
       "       'y_translation': None,\n",
       "       'z_rotation': None,\n",
       "       'z_scaling': None,\n",
       "       'z_translation': None},\n",
       "      'x_dimension': None,\n",
       "      'y_dimension': None,\n",
       "      'z_dimension': None,\n",
       "      'all_collisions': [{'@type': None,\n",
       "        'collision_method': None,\n",
       "        'collisions': [{'@type': None,\n",
       "          'reference_organ': None,\n",
       "          'as_3d_id': None,\n",
       "          'as_id': None,\n",
       "          'as_label': None,\n",
       "          'as_volume': None,\n",
       "          'percentage': None}]}],\n",
       "      'corridor': {'@type': None, 'file_format': None, 'file': None},\n",
       "      'summaries': [{'@type': None,\n",
       "        'sex': None,\n",
       "        'annotation_method': None,\n",
       "        'aggregated_summary_count': None,\n",
       "        'aggregated_summaries': [None],\n",
       "        'modality': None,\n",
       "        'summary': [{'@type': None,\n",
       "          'cell_id': None,\n",
       "          'cell_label': None,\n",
       "          'count': None,\n",
       "          'percentage': None}]}]}}],\n",
       "   'consortium_name': None,\n",
       "   'provider_name': None,\n",
       "   'provider_uuid': None,\n",
       "   'age': None,\n",
       "   'sex': None,\n",
       "   'bmi': None,\n",
       "   'race': None,\n",
       "   'link': None,\n",
       "   'label': None,\n",
       "   'description': None}]}"
      ]
     },
     "execution_count": 9,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "json_structure"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 31,
   "id": "f3e91c01",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "      cell_id            cell_label annotation_method            modality  \\\n",
      "0  CL:0009080  intestinal tuft cell              popv  sc_transcriptomics   \n",
      "1  CL:0000057            fibroblast              popv  sc_transcriptomics   \n",
      "2  CL:0000775            neutrophil              popv  sc_transcriptomics   \n",
      "3  CL:0000775            neutrophil              popv  sc_transcriptomics   \n",
      "4  CL:0000097             mast cell              popv  sc_transcriptomics   \n",
      "\n",
      "  cell_source_label     sex  \\\n",
      "0           jejunum  Female   \n",
      "1           jejunum  Female   \n",
      "2           jejunum  Female   \n",
      "3           jejunum  Female   \n",
      "4           jejunum  Female   \n",
      "\n",
      "                                aggregated_summaries  #datasets  \n",
      "0  [https://entity.api.hubmapconsortium.org/entit...          2  \n",
      "1  [https://entity.api.hubmapconsortium.org/entit...          2  \n",
      "2  [https://entity.api.hubmapconsortium.org/entit...          2  \n",
      "3  [https://entity.api.hubmapconsortium.org/entit...          2  \n",
      "4  [https://entity.api.hubmapconsortium.org/entit...          2   \n",
      "\n",
      "Final CSV saved to: output/CTs-with-datasets.csv\n"
     ]
    },
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Supriya\\AppData\\Local\\Temp\\ipykernel_11004\\3555604534.py:30: FutureWarning: Series.fillna with 'method' is deprecated and will raise in a future version. Use obj.ffill() or obj.bfill() instead.\n",
      "  merged_json_df['modality'] = merged_json_df['modality'].fillna(method='ffill')\n",
      "C:\\Users\\Supriya\\AppData\\Local\\Temp\\ipykernel_11004\\3555604534.py:67: SettingWithCopyWarning: \n",
      "A value is trying to be set on a copy of a slice from a DataFrame.\n",
      "Try using .loc[row_indexer,col_indexer] = value instead\n",
      "\n",
      "See the caveats in the documentation: https://pandas.pydata.org/pandas-docs/stable/user_guide/indexing.html#returning-a-view-versus-a-copy\n",
      "  final_df['#datasets'] = final_df['aggregated_summaries'].apply(lambda x: len(x) if isinstance(x, list) else 0)\n"
     ]
    }
   ],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Load the provided CSV and JSON files\n",
    "ftu_file_path = 'ref_data\\\\FTU Cell Count Table - Cell_Type_Count.csv'\n",
    "json_file_path = 'ref_data\\\\atlas-as-cell-summaries.jsonld'\n",
    "\n",
    "# Read the CSV file\n",
    "ftu_data = pd.read_csv(ftu_file_path)\n",
    "ftu_data = ftu_data.drop(columns=['Unnamed: 6','Unnamed: 7','Unnamed: 8','Unnamed: 9','Unnamed: 10','Unnamed: 11','Unnamed: 12','Unnamed: 13','Unnamed: 14'], axis = False)\n",
    "\n",
    "# Load the JSON file\n",
    "with open(json_file_path, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "# Extract the relevant data from the JSON file\n",
    "json_graph = json_data['@graph']\n",
    "\n",
    "# Convert the JSON graph to a DataFrame for easier manipulation\n",
    "json_df = pd.json_normalize(json_graph)\n",
    "\n",
    "# Ensure modality is expanded correctly\n",
    "expanded_json_df = json_df.explode('summary').reset_index(drop=True)\n",
    "expanded_json_df['modality'] = json_df.explode('modality').reset_index(drop=True)['modality']\n",
    "\n",
    "# Normalize the summary column and merge with expanded_json_df\n",
    "summary_df = pd.json_normalize(expanded_json_df['summary'])\n",
    "merged_json_df = pd.concat([expanded_json_df.drop(columns=['summary']), summary_df], axis=1)\n",
    "\n",
    "# Ensure correct modality assignment\n",
    "merged_json_df['modality'] = merged_json_df['modality'].fillna(method='ffill')\n",
    "\n",
    "# Extract relevant columns from the CSV data\n",
    "csv_relevant_columns = csv_data[['CT ID in CL', 'Organ', 'FTU Label in Uberon', 'FTU ID in Uberon', 'CT Label in CL']]\n",
    "\n",
    "# Merge the CSV and JSON DataFrames based on matching cell_id\n",
    "merged_data = pd.merge(\n",
    "    merged_json_df,\n",
    "    csv_relevant_columns,\n",
    "    left_on='cell_id',\n",
    "    right_on='CT ID in CL',\n",
    "    how='inner'\n",
    ")\n",
    "\n",
    "# Create the final DataFrame with the desired columns\n",
    "final_df = merged_data[[\n",
    "    'cell_id',\n",
    "    'cell_label',\n",
    "    'annotation_method',\n",
    "    'modality',\n",
    "    'cell_source_label',\n",
    "    'sex',\n",
    "    'aggregated_summaries'\n",
    "]]\n",
    "\n",
    "# Rename the columns as specified\n",
    "final_df.columns = [\n",
    "    'cell_id',\n",
    "    'cell_label',\n",
    "    'annotation_method',\n",
    "    'modality',\n",
    "    'cell_source_label',\n",
    "    'sex',\n",
    "    'aggregated_summaries'\n",
    "]\n",
    "\n",
    "# Add the new column \"#datasets\"\n",
    "final_df['#datasets'] = final_df['aggregated_summaries'].apply(lambda x: len(x) if isinstance(x, list) else 0)\n",
    "\n",
    "# Check for null values in the 'modality' column again\n",
    "null_modality_entries_after_correction = final_df[final_df['modality'].isnull()]\n",
    "\n",
    "# Save the final DataFrame to a CSV file\n",
    "final_csv_path = 'output/CTs-with-datasets.csv'\n",
    "final_df.to_csv(final_csv_path, index=False)\n",
    "\n",
    "# Display the first few rows of the final DataFrame\n",
    "print(final_df.head(), '\\n')\n",
    "\n",
    "print(f\"Final CSV saved to: {final_csv_path}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 11,
   "id": "5019c964",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['sc_transcriptomics'], dtype=object)"
      ]
     },
     "execution_count": 11,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df['modality'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 55,
   "id": "f3ffd8c8",
   "metadata": {},
   "outputs": [],
   "source": [
    "import pandas as pd\n",
    "import json\n",
    "\n",
    "# Load CSV file\n",
    "csv_file_path = 'output/CTs-with-datasets.csv'\n",
    "csv_df = pd.read_csv(csv_file_path)\n",
    "\n",
    "# Load JSON file\n",
    "json_file_path = 'ref_data/atlas-enriched-dataset-graph.jsonld'\n",
    "with open(json_file_path, 'r') as f:\n",
    "    json_data = json.load(f)\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 78,
   "id": "88ab5535",
   "metadata": {},
   "outputs": [
    {
     "name": "stderr",
     "output_type": "stream",
     "text": [
      "C:\\Users\\Supriya\\AppData\\Local\\Temp\\ipykernel_11004\\1379520534.py:61: FutureWarning: The behavior of DataFrame concatenation with empty or all-NA entries is deprecated. In a future version, this will no longer exclude empty or all-NA columns when determining the result dtypes. To retain the old behavior, exclude the relevant entries before the concat operation.\n",
      "  result_df = pd.concat([result_df, pd.DataFrame(matches)], ignore_index=True)\n"
     ]
    },
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "                                        organ_id organ_name  \\\n",
      "0  http://purl.obolibrary.org/obo/UBERON_0002108    unknown   \n",
      "1  http://purl.obolibrary.org/obo/UBERON_0002108    unknown   \n",
      "2  http://purl.obolibrary.org/obo/UBERON_0002108    unknown   \n",
      "3  http://purl.obolibrary.org/obo/UBERON_0002108    unknown   \n",
      "4  http://purl.obolibrary.org/obo/UBERON_0002108    unknown   \n",
      "\n",
      "                                     reference_organ     cell_id  \\\n",
      "0  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
      "1  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
      "2  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
      "3  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
      "4  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
      "\n",
      "             cell_label annotation_method cell_source_label as_label     sex  \\\n",
      "0  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
      "1  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
      "2  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
      "3  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
      "4  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
      "\n",
      "                                aggregated_summaries           gene_id  \\\n",
      "0  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:plcg2   \n",
      "1  https://entity.api.hubmapconsortium.org/entiti...   ASCTB-TEMP:st18   \n",
      "2  https://entity.api.hubmapconsortium.org/entiti...   ASCTB-TEMP:grk5   \n",
      "3  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:itpr2   \n",
      "4  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:zfhx3   \n",
      "\n",
      "  gene_label ensembl_id  mean_gene_expr_value  \n",
      "0      PLCG2      PLCG2              4.136788  \n",
      "1       ST18       ST18              3.566198  \n",
      "2       GRK5       GRK5              3.731159  \n",
      "3      ITPR2      ITPR2              4.091242  \n",
      "4      ZFHX3      ZFHX3              3.002826  \n"
     ]
    }
   ],
   "source": [
    "# Create an empty dataframe with the specified columns\n",
    "columns = [\n",
    "    'organ_id', 'organ_name', 'reference_organ', 'cell_id', 'cell_label', 'annotation_method', \n",
    "    'cell_source_label', 'as_label', 'sex', 'aggregated_summaries', 'gene_id', \n",
    "    'gene_label', 'ensembl_id', 'mean_gene_expr_value'\n",
    "]\n",
    "result_df = pd.DataFrame(columns=columns)\n",
    "\n",
    "# Function to extract and match data based on cell_id and annotation_method\n",
    "def extract_data(csv_row, json_data):\n",
    "    matches = []\n",
    "    cell_id = csv_row['cell_id']\n",
    "    annotation_method = csv_row['annotation_method']\n",
    "    aggregated_summaries = eval(csv_row['aggregated_summaries'])  # convert string representation of list to actual list\n",
    "    organ_name = 'unknown'\n",
    "    for entry in json_data[\"@graph\"]:\n",
    "        for sample in entry.get('samples', []):\n",
    "            rui_location = sample.get('rui_location', {})\n",
    "            all_collisions = rui_location.get('all_collisions', [])\n",
    "            for collision in all_collisions:\n",
    "                collisions = collision.get('collisions', [])\n",
    "                for col in collisions:\n",
    "                    reference_organ = col.get('reference_organ', 'unknown')\n",
    "                    as_label = col.get('as_label', 'unknown')\n",
    "                    for section in sample.get('sections', []):\n",
    "                        for dataset in section.get('datasets', []):\n",
    "                            summaries = dataset.get('summaries', [])\n",
    "                            organ_id = dataset.get('organ_id', 'unknown')\n",
    "                            for summary in summaries:\n",
    "                                if summary.get('annotation_method') == annotation_method:\n",
    "                                    for summ in summary['summary']:\n",
    "                                        if isinstance(summ, dict) and summ['cell_id'] == cell_id:\n",
    "                                            for aggregated_summary in aggregated_summaries:\n",
    "                                                if aggregated_summary in dataset['@id']:\n",
    "                                                    for gene_expr in summ.get('gene_expr', []):\n",
    "                                                        if isinstance(gene_expr, dict):  # Ensure gene_expr is a dictionary\n",
    "                                                            new_row = {\n",
    "                                                                'organ_id': organ_id,\n",
    "                                                                'organ_name': organ_name,\n",
    "                                                                'cell_id': cell_id,\n",
    "                                                                'cell_label': summ['cell_label'],\n",
    "                                                                'annotation_method': annotation_method,\n",
    "                                                                'cell_source_label': csv_row['cell_source_label'],\n",
    "                                                                'sex': csv_row['sex'],\n",
    "                                                                'aggregated_summaries': aggregated_summary,\n",
    "                                                                'gene_id': gene_expr.get('gene_id'),\n",
    "                                                                'gene_label': gene_expr.get('gene_label'),\n",
    "                                                                'ensembl_id': gene_expr.get('ensembl_id'),\n",
    "                                                                'mean_gene_expr_value': gene_expr.get('mean_gene_expr_value'),\n",
    "                                                                'reference_organ': reference_organ,\n",
    "                                                                'as_label': as_label\n",
    "                                                            }\n",
    "                                                            matches.append(new_row)\n",
    "    return matches\n",
    "\n",
    "# Iterate over the CSV data and extract matched data\n",
    "for index, row in csv_data.iterrows():\n",
    "    if row['cell_source_label']:  # Filter data by cell_source_label\n",
    "        matches = extract_data(row, json_data)\n",
    "        if matches:\n",
    "            result_df = pd.concat([result_df, pd.DataFrame(matches)], ignore_index=True)\n",
    "\n",
    "# Display the new dataframe\n",
    "print(result_df.head())\n"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 85,
   "id": "445fdc6f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['http://purl.obolibrary.org/obo/UBERON_0002108',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0000059',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0002113',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0000948',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0001255',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0002048',\n",
       "       'http://purl.obolibrary.org/obo/UBERON_0002107'], dtype=object)"
      ]
     },
     "execution_count": 85,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Get unique organ IDs\n",
    "result_df[\"organ_id\"].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 86,
   "id": "d0dec88d",
   "metadata": {},
   "outputs": [],
   "source": [
    "organ_id_to_name = {\n",
    "    'http://purl.obolibrary.org/obo/UBERON_0002108' : 'small intestine',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0000059': 'large intestine',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0002113': 'kidney',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0000948': 'heart',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0001255': 'urinary bladder',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0002048': 'lung',\n",
    "'http://purl.obolibrary.org/obo/UBERON_0002107': 'liver'}"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 87,
   "id": "d8fd22c5",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_id</th>\n",
       "      <th>organ_name</th>\n",
       "      <th>reference_organ</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>annotation_method</th>\n",
       "      <th>cell_source_label</th>\n",
       "      <th>as_label</th>\n",
       "      <th>sex</th>\n",
       "      <th>aggregated_summaries</th>\n",
       "      <th>gene_id</th>\n",
       "      <th>gene_label</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>mean_gene_expr_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002108</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...</td>\n",
       "      <td>CL:0009080</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>popv</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>Female</td>\n",
       "      <td>https://entity.api.hubmapconsortium.org/entiti...</td>\n",
       "      <td>ASCTB-TEMP:plcg2</td>\n",
       "      <td>PLCG2</td>\n",
       "      <td>PLCG2</td>\n",
       "      <td>4.136788</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002108</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...</td>\n",
       "      <td>CL:0009080</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>popv</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>Female</td>\n",
       "      <td>https://entity.api.hubmapconsortium.org/entiti...</td>\n",
       "      <td>ASCTB-TEMP:st18</td>\n",
       "      <td>ST18</td>\n",
       "      <td>ST18</td>\n",
       "      <td>3.566198</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002108</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...</td>\n",
       "      <td>CL:0009080</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>popv</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>Female</td>\n",
       "      <td>https://entity.api.hubmapconsortium.org/entiti...</td>\n",
       "      <td>ASCTB-TEMP:grk5</td>\n",
       "      <td>GRK5</td>\n",
       "      <td>GRK5</td>\n",
       "      <td>3.731159</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002108</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...</td>\n",
       "      <td>CL:0009080</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>popv</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>Female</td>\n",
       "      <td>https://entity.api.hubmapconsortium.org/entiti...</td>\n",
       "      <td>ASCTB-TEMP:itpr2</td>\n",
       "      <td>ITPR2</td>\n",
       "      <td>ITPR2</td>\n",
       "      <td>4.091242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002108</td>\n",
       "      <td>small intestine</td>\n",
       "      <td>http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...</td>\n",
       "      <td>CL:0009080</td>\n",
       "      <td>intestinal tuft cell</td>\n",
       "      <td>popv</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>jejunum</td>\n",
       "      <td>Female</td>\n",
       "      <td>https://entity.api.hubmapconsortium.org/entiti...</td>\n",
       "      <td>ASCTB-TEMP:zfhx3</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>ZFHX3</td>\n",
       "      <td>3.002826</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "                                        organ_id       organ_name  \\\n",
       "0  http://purl.obolibrary.org/obo/UBERON_0002108  small intestine   \n",
       "1  http://purl.obolibrary.org/obo/UBERON_0002108  small intestine   \n",
       "2  http://purl.obolibrary.org/obo/UBERON_0002108  small intestine   \n",
       "3  http://purl.obolibrary.org/obo/UBERON_0002108  small intestine   \n",
       "4  http://purl.obolibrary.org/obo/UBERON_0002108  small intestine   \n",
       "\n",
       "                                     reference_organ     cell_id  \\\n",
       "0  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
       "1  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
       "2  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
       "3  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
       "4  http://purl.org/ccf/latest/ccf.owl#VHFSmallInt...  CL:0009080   \n",
       "\n",
       "             cell_label annotation_method cell_source_label as_label     sex  \\\n",
       "0  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
       "1  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
       "2  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
       "3  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
       "4  intestinal tuft cell              popv           jejunum  jejunum  Female   \n",
       "\n",
       "                                aggregated_summaries           gene_id  \\\n",
       "0  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:plcg2   \n",
       "1  https://entity.api.hubmapconsortium.org/entiti...   ASCTB-TEMP:st18   \n",
       "2  https://entity.api.hubmapconsortium.org/entiti...   ASCTB-TEMP:grk5   \n",
       "3  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:itpr2   \n",
       "4  https://entity.api.hubmapconsortium.org/entiti...  ASCTB-TEMP:zfhx3   \n",
       "\n",
       "  gene_label ensembl_id  mean_gene_expr_value  \n",
       "0      PLCG2      PLCG2              4.136788  \n",
       "1       ST18       ST18              3.566198  \n",
       "2       GRK5       GRK5              3.731159  \n",
       "3      ITPR2      ITPR2              4.091242  \n",
       "4      ZFHX3      ZFHX3              3.002826  "
      ]
     },
     "execution_count": 87,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adding a new column to the dataframe\n",
    "result_df['organ_name'] = result_df['organ_id'].map(organ_id_to_name)\n",
    "\n",
    "# Display the first few rows of the updated dataframe\n",
    "result_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 89,
   "id": "b44a6a13",
   "metadata": {},
   "outputs": [],
   "source": [
    "result_df.to_csv('output/filtered-CTs-with-datasets-with-gene-information_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "54b46188",
   "metadata": {},
   "source": [
    "##### Sanity check of the organ_id, cell_source_label and cell_ids"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 90,
   "id": "51d587b2",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['jejunum', 'descending colon', 'superior part of duodenum',\n",
       "       'descending part of duodenum', 'ascending part of duodenum',\n",
       "       'horizontal part of duodenum', 'sigmoid colon',\n",
       "       'distal part of ileum', 'ileum', 'renal pyramid',\n",
       "       'outer cortex of kidney', 'heart left ventricle', 'rectum',\n",
       "       'ascending colon', 'transverse colon', 'renal column',\n",
       "       'renal papilla', 'fundus of urinary bladder',\n",
       "       'Right Medial Bronchopulmonary Segment',\n",
       "       'Right Lateral Bronchopulmonary Segment', 'left ureter',\n",
       "       'hilum of kidney', 'kidney capsule', 'gastric impression of liver',\n",
       "       'diaphragmatic surface of liver', 'capsule of the liver'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 90,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "result_df['cell_source_label'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 91,
   "id": "13cfb719",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['jejunum', 'descending colon', 'superior part of duodenum',\n",
       "       'descending part of duodenum', 'ascending part of duodenum',\n",
       "       'horizontal part of duodenum', 'sigmoid colon',\n",
       "       'distal part of ileum', 'ileum', 'renal pyramid',\n",
       "       'outer cortex of kidney', 'diaphragmatic surface of spleen',\n",
       "       'hilum of spleen', 'heart right ventricle',\n",
       "       'interventricular septum', 'heart left ventricle', 'rectum',\n",
       "       'ascending colon', 'transverse colon', 'renal column',\n",
       "       'renal papilla', 'fundus of urinary bladder',\n",
       "       'Right Medial Bronchopulmonary Segment',\n",
       "       'Right Lateral Bronchopulmonary Segment', 'left ureter',\n",
       "       'hilum of kidney', 'kidney capsule', 'caecum',\n",
       "       'gastric impression of liver', 'diaphragmatic surface of liver',\n",
       "       'capsule of the liver', 'left cardiac atrium',\n",
       "       'right cardiac atrium',\n",
       "       'Posteromedial head of posterior papillary muscle of left ventricle',\n",
       "       'Lateral segmental bronchus',\n",
       "       'Left apical bronchopulmonary segment', 'central zone of prostate',\n",
       "       'peripheral zone of prostate', 'skin of body',\n",
       "       'Interlobar adipose tissue of right mammary gland'], dtype=object)"
      ]
     },
     "execution_count": 91,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_df['cell_source_label'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 92,
   "id": "e19c2280",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Difference in cell_source_label:\n",
      "{'right cardiac atrium', 'peripheral zone of prostate', 'Posteromedial head of posterior papillary muscle of left ventricle', 'caecum', 'hilum of spleen', 'left cardiac atrium', 'Interlobar adipose tissue of right mammary gland', 'skin of body', 'Left apical bronchopulmonary segment', 'central zone of prostate', 'diaphragmatic surface of spleen', 'interventricular septum', 'heart right ventricle', 'Lateral segmental bronchus'}\n"
     ]
    }
   ],
   "source": [
    "difference_labels = set(final_df['cell_source_label'].unique()) - set(result_df['cell_source_label'])\n",
    "print(\"Difference in cell_source_label:\")\n",
    "print(difference_labels)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 96,
   "id": "155a6a96",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Difference in cell_ids:\n",
      "{'CL:0000788', 'CL:1000458', 'CL:0000071', 'CL:0009016', 'CL:0002341', 'CL:0002457', 'CL:0000787', 'CL:0002340', 'CL:0001065', 'CL:0000499'}\n"
     ]
    }
   ],
   "source": [
    "difference_ids = set(final_df['cell_id'].unique()) - set(result_df['cell_id'])\n",
    "print(\"Difference in cell_ids:\")\n",
    "print(difference_ids)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 97,
   "id": "1aafb7f6",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Find the unique cell_labels in the provided CSV file\n",
    "ftu_cell_labels = set(ftu_data['CT Label in CL'].unique())\n",
    "ftu_cell_ids = set(ftu_data['CT ID in CL'].unique())"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 101,
   "id": "dc073f2f",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'B cell',\n",
       " 'basal cell',\n",
       " 'basal cell of prostate epithelium',\n",
       " 'blood vessel endothelial cell',\n",
       " 'capillary endothelial cell',\n",
       " 'dendritic cell',\n",
       " 'endothelial cell',\n",
       " 'endothelial cell of artery',\n",
       " 'enterocyte of epithelium of large intestine',\n",
       " 'fibroblast',\n",
       " 'hepatocyte',\n",
       " 'innate lymphoid cell',\n",
       " 'intestinal crypt stem cell of large intestine',\n",
       " 'intestinal tuft cell',\n",
       " 'luminal cell of prostate epithelium',\n",
       " 'macrophage',\n",
       " 'mast cell',\n",
       " 'memory B cell',\n",
       " 'naive B cell',\n",
       " 'neutrophil',\n",
       " 'pericyte',\n",
       " 'plasma cell',\n",
       " 'smooth muscle cell',\n",
       " 'stromal cell',\n",
       " 'type I pneumocyte',\n",
       " 'type II pneumocyte',\n",
       " 'vascular associated smooth muscle cell',\n",
       " 'vein endothelial cell'}"
      ]
     },
     "execution_count": 101,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_labels_set = set()\n",
    "for label in difference_labels:\n",
    "    cell_labels = final_df[final_df['cell_source_label'] == label]['cell_label']\n",
    "    cell_labels_set.update(cell_labels) \n",
    "    \n",
    "# Check for matches between cell_labels_set and csv_cell_labels\n",
    "matches_label = cell_labels_set.intersection(ftu_cell_labels)\n",
    "matches_label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 102,
   "id": "076e500b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "{'CL:0000071',\n",
       " 'CL:0000499',\n",
       " 'CL:0000787',\n",
       " 'CL:0000788',\n",
       " 'CL:0001065',\n",
       " 'CL:0002340',\n",
       " 'CL:0002341',\n",
       " 'CL:0002457',\n",
       " 'CL:0009016',\n",
       " 'CL:1000458'}"
      ]
     },
     "execution_count": 102,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "cell_ids_set = set()\n",
    "for ids in difference_ids:\n",
    "    cell_ids = final_df[final_df['cell_id'] == ids]['cell_id']\n",
    "    cell_ids_set.update(cell_ids)\n",
    "    \n",
    "matches_id = cell_ids_set.intersection(ftu_cell_ids)\n",
    "matches_id"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 105,
   "id": "ed2922ef",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>Organ</th>\n",
       "      <th>FTU Label in Uberon</th>\n",
       "      <th>FTU ID in Uberon</th>\n",
       "      <th>CT Label in CL</th>\n",
       "      <th>CT ID in CL</th>\n",
       "      <th>CT Label in 2D Object</th>\n",
       "      <th>Unnamed: 6</th>\n",
       "      <th>Unnamed: 7</th>\n",
       "      <th>Unnamed: 8</th>\n",
       "      <th>Unnamed: 9</th>\n",
       "      <th>Unnamed: 10</th>\n",
       "      <th>Unnamed: 11</th>\n",
       "      <th>Unnamed: 12</th>\n",
       "      <th>Unnamed: 13</th>\n",
       "      <th>Unnamed: 14</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>Thymus Lobule</td>\n",
       "      <td>UBERON:0002125</td>\n",
       "      <td>naive B cell</td>\n",
       "      <td>CL:0000788</td>\n",
       "      <td>B Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Skin</td>\n",
       "      <td>Epidermal Ridge</td>\n",
       "      <td>UBERON:0013487</td>\n",
       "      <td>melanocyte of skin</td>\n",
       "      <td>CL:1000458</td>\n",
       "      <td>Melanocyte</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>Thymus Lobule</td>\n",
       "      <td>UBERON:0002125</td>\n",
       "      <td>blood vessel endothelial cell</td>\n",
       "      <td>CL:0000071</td>\n",
       "      <td>Blood Vessel Endothelial Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Liver</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Large Intestine</td>\n",
       "      <td>Crypt of Lieberkuhn</td>\n",
       "      <td>UBERON:0001984</td>\n",
       "      <td>intestinal crypt stem cell of large intestine</td>\n",
       "      <td>CL:0009016</td>\n",
       "      <td>Epithelial Stem Cells</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>Prostate Glandular Acinus</td>\n",
       "      <td>UBERON:0004179</td>\n",
       "      <td>basal cell of prostate epithelium</td>\n",
       "      <td>CL:0002341</td>\n",
       "      <td>Basal Cell Of Prostate Epithelium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>5</th>\n",
       "      <td>Skin</td>\n",
       "      <td>Epidermal Ridge</td>\n",
       "      <td>UBERON:0013487</td>\n",
       "      <td>epidermal Langerhans cell</td>\n",
       "      <td>CL:0002457</td>\n",
       "      <td>Epidermal Langerhans Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>6</th>\n",
       "      <td>Thymus</td>\n",
       "      <td>Thymus Lobule</td>\n",
       "      <td>UBERON:0002125</td>\n",
       "      <td>memory B cell</td>\n",
       "      <td>CL:0000787</td>\n",
       "      <td>B Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>Prostate</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>7</th>\n",
       "      <td>Prostate</td>\n",
       "      <td>Prostate Glandular Acinus</td>\n",
       "      <td>UBERON:0004179</td>\n",
       "      <td>luminal cell of prostate epithelium</td>\n",
       "      <td>CL:0002340</td>\n",
       "      <td>Luminal Cell Of Prostate Epithelium</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>8</th>\n",
       "      <td>Small Intestine</td>\n",
       "      <td>intestinal villus</td>\n",
       "      <td>UBERON 0001213</td>\n",
       "      <td>innate lymphoid cell</td>\n",
       "      <td>CL:0001065</td>\n",
       "      <td>innate lymphoid cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>9</th>\n",
       "      <td>Spleen</td>\n",
       "      <td>White Pulp</td>\n",
       "      <td>UBERON:0001959</td>\n",
       "      <td>stromal cell</td>\n",
       "      <td>CL:0000499</td>\n",
       "      <td>Stromal Cell</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "      <td>NaN</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "             Organ        FTU Label in Uberon FTU ID in Uberon  \\\n",
       "0           Thymus              Thymus Lobule   UBERON:0002125   \n",
       "1             Skin            Epidermal Ridge   UBERON:0013487   \n",
       "2           Thymus              Thymus Lobule   UBERON:0002125   \n",
       "3  Large Intestine        Crypt of Lieberkuhn   UBERON:0001984   \n",
       "4         Prostate  Prostate Glandular Acinus   UBERON:0004179   \n",
       "5             Skin            Epidermal Ridge   UBERON:0013487   \n",
       "6           Thymus              Thymus Lobule   UBERON:0002125   \n",
       "7         Prostate  Prostate Glandular Acinus   UBERON:0004179   \n",
       "8  Small Intestine          intestinal villus   UBERON 0001213   \n",
       "9           Spleen                 White Pulp   UBERON:0001959   \n",
       "\n",
       "                                  CT Label in CL CT ID in CL  \\\n",
       "0                                   naive B cell  CL:0000788   \n",
       "1                             melanocyte of skin  CL:1000458   \n",
       "2                  blood vessel endothelial cell  CL:0000071   \n",
       "3  intestinal crypt stem cell of large intestine  CL:0009016   \n",
       "4              basal cell of prostate epithelium  CL:0002341   \n",
       "5                      epidermal Langerhans cell  CL:0002457   \n",
       "6                                  memory B cell  CL:0000787   \n",
       "7            luminal cell of prostate epithelium  CL:0002340   \n",
       "8                           innate lymphoid cell  CL:0001065   \n",
       "9                                   stromal cell  CL:0000499   \n",
       "\n",
       "                 CT Label in 2D Object  Unnamed: 6  Unnamed: 7  Unnamed: 8  \\\n",
       "0                               B Cell         NaN         NaN         NaN   \n",
       "1                           Melanocyte         NaN         NaN         NaN   \n",
       "2        Blood Vessel Endothelial Cell         NaN         NaN         NaN   \n",
       "3                Epithelial Stem Cells         NaN         NaN         NaN   \n",
       "4    Basal Cell Of Prostate Epithelium         NaN         NaN         NaN   \n",
       "5            Epidermal Langerhans Cell         NaN         NaN         NaN   \n",
       "6                               B Cell         NaN         NaN         NaN   \n",
       "7  Luminal Cell Of Prostate Epithelium         NaN         NaN         NaN   \n",
       "8                 innate lymphoid cell         NaN         NaN         NaN   \n",
       "9                         Stromal Cell         NaN         NaN         NaN   \n",
       "\n",
       "   Unnamed: 9  Unnamed: 10  Unnamed: 11  Unnamed: 12 Unnamed: 13 Unnamed: 14  \n",
       "0         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "1         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "2         NaN          NaN          NaN          NaN         NaN       Liver  \n",
       "3         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "4         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "5         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "6         NaN          NaN          NaN          NaN         NaN    Prostate  \n",
       "7         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "8         NaN          NaN          NaN          NaN         NaN         NaN  \n",
       "9         NaN          NaN          NaN          NaN         NaN         NaN  "
      ]
     },
     "execution_count": 105,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "match_df = pd.DataFrame()\n",
    "for ids in matches_id:\n",
    "    match_df = pd.concat([match_df, (ftu_data[ftu_data['CT ID in CL'] == ids])],ignore_index=True)\n",
    "\n",
    "match_df"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 40,
   "id": "72b870ab",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['diaphragmatic surface of spleen', 'hilum of spleen',\n",
       "       'central zone of prostate', 'peripheral zone of prostate',\n",
       "       'skin of body', 'descending colon', 'sigmoid colon', 'rectum',\n",
       "       'ascending colon', 'transverse colon', 'caecum', 'jejunum',\n",
       "       'superior part of duodenum', 'descending part of duodenum',\n",
       "       'ascending part of duodenum', 'horizontal part of duodenum',\n",
       "       'distal part of ileum', 'ileum', 'fundus of urinary bladder',\n",
       "       'Interlobar adipose tissue of right mammary gland',\n",
       "       'Right Medial Bronchopulmonary Segment',\n",
       "       'Right Lateral Bronchopulmonary Segment',\n",
       "       'gastric impression of liver', 'diaphragmatic surface of liver',\n",
       "       'capsule of the liver', 'Lateral segmental bronchus',\n",
       "       'Left apical bronchopulmonary segment', 'heart right ventricle',\n",
       "       'interventricular septum', 'heart left ventricle',\n",
       "       'left cardiac atrium', 'right cardiac atrium',\n",
       "       'Posteromedial head of posterior papillary muscle of left ventricle'],\n",
       "      dtype=object)"
      ]
     },
     "execution_count": 40,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "matched_data = pd.DataFrame()\n",
    "for cl_label in matches:\n",
    "    matched_data = pd.concat([matched_data, final_df[final_df['cell_label'] == cl_label]])\n",
    "    \n",
    "matched_data['cell_source_label'].unique()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "0d2a7b9b",
   "metadata": {},
   "source": [
    "### Organ information from CTann crosswalk files"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 60,
   "id": "aaa4754d",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>CL_ID</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>annotation_method</th>\n",
       "      <th>modality</th>\n",
       "      <th>cell_source_label</th>\n",
       "      <th>sex</th>\n",
       "      <th>aggregated_summaries</th>\n",
       "      <th>#datasets</th>\n",
       "      <th>Organ_Level</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL:1000716</td>\n",
       "      <td>Outer Medullary Collecting Duct Principal</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>sc_transcriptomics</td>\n",
       "      <td>renal pyramid</td>\n",
       "      <td>Female</td>\n",
       "      <td>['https://entity.api.hubmapconsortium.org/enti...</td>\n",
       "      <td>55</td>\n",
       "      <td>Kidney_L3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL:1000716</td>\n",
       "      <td>Outer Medullary Collecting Duct Principal</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>sc_transcriptomics</td>\n",
       "      <td>renal pyramid</td>\n",
       "      <td>Female</td>\n",
       "      <td>['https://entity.api.hubmapconsortium.org/enti...</td>\n",
       "      <td>55</td>\n",
       "      <td>Kidney_L2</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CL:1000768</td>\n",
       "      <td>Connecting Tubule</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>sc_transcriptomics</td>\n",
       "      <td>renal pyramid</td>\n",
       "      <td>Female</td>\n",
       "      <td>['https://entity.api.hubmapconsortium.org/enti...</td>\n",
       "      <td>55</td>\n",
       "      <td>Kidney_L3</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CL:1000768</td>\n",
       "      <td>Connecting Tubule</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>sc_transcriptomics</td>\n",
       "      <td>renal pyramid</td>\n",
       "      <td>Female</td>\n",
       "      <td>['https://entity.api.hubmapconsortium.org/enti...</td>\n",
       "      <td>55</td>\n",
       "      <td>Kidney_L1</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CL:1000768</td>\n",
       "      <td>Connecting Tubule</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>sc_transcriptomics</td>\n",
       "      <td>renal pyramid</td>\n",
       "      <td>Female</td>\n",
       "      <td>['https://entity.api.hubmapconsortium.org/enti...</td>\n",
       "      <td>55</td>\n",
       "      <td>Kidney_L2</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "        CL_ID                                 cell_label annotation_method  \\\n",
       "0  CL:1000716  Outer Medullary Collecting Duct Principal           azimuth   \n",
       "1  CL:1000716  Outer Medullary Collecting Duct Principal           azimuth   \n",
       "2  CL:1000768                          Connecting Tubule           azimuth   \n",
       "3  CL:1000768                          Connecting Tubule           azimuth   \n",
       "4  CL:1000768                          Connecting Tubule           azimuth   \n",
       "\n",
       "             modality cell_source_label     sex  \\\n",
       "0  sc_transcriptomics     renal pyramid  Female   \n",
       "1  sc_transcriptomics     renal pyramid  Female   \n",
       "2  sc_transcriptomics     renal pyramid  Female   \n",
       "3  sc_transcriptomics     renal pyramid  Female   \n",
       "4  sc_transcriptomics     renal pyramid  Female   \n",
       "\n",
       "                                aggregated_summaries  #datasets Organ_Level  \n",
       "0  ['https://entity.api.hubmapconsortium.org/enti...         55   Kidney_L3  \n",
       "1  ['https://entity.api.hubmapconsortium.org/enti...         55   Kidney_L2  \n",
       "2  ['https://entity.api.hubmapconsortium.org/enti...         55   Kidney_L3  \n",
       "3  ['https://entity.api.hubmapconsortium.org/enti...         55   Kidney_L1  \n",
       "4  ['https://entity.api.hubmapconsortium.org/enti...         55   Kidney_L2  "
      ]
     },
     "execution_count": 60,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the uploaded files\n",
    "final_output_df = pd.read_csv('output/CTs-with-datasets.csv')\n",
    "popv_df = pd.read_csv('ref_data/popv.csv')\n",
    "celltypist_df = pd.read_csv('ref_data/celltypist.csv')\n",
    "azimuth_df = pd.read_csv('ref_data/azimuth.csv')\n",
    "\n",
    "# Rename the column in final_output_df\n",
    "final_output_df.rename(columns={'cell_id': 'CL_ID'}, inplace=True)\n",
    "\n",
    "# Merge with azimuth_df\n",
    "final_output_azimuth = final_output_df[final_output_df['annotation_method'] == 'azimuth'].merge(\n",
    "    azimuth_df[['CL_ID', 'Organ_Level']], on='CL_ID', how='left')\n",
    "\n",
    "# Merge with celltypist_df\n",
    "final_output_celltypist = final_output_df[final_output_df['annotation_method'] == 'celltypist'].merge(\n",
    "    celltypist_df[['CL_ID', 'Organ_Level']], on='CL_ID', how='left')\n",
    "\n",
    "# Merge with popv_df\n",
    "final_output_popv = final_output_df[final_output_df['annotation_method'] == 'popv'].merge(\n",
    "    popv_df[['CL_ID', 'Organ_Level']], on='CL_ID', how='left')\n",
    "\n",
    "# Combine all the dataframes\n",
    "final_output_combined = pd.concat([final_output_azimuth, final_output_celltypist, final_output_popv], ignore_index=True)\n",
    "\n",
    "# Define the organ levels for each annotation method\n",
    "azimuth_organs = ['Kidney_L3', 'Lung_v2_finest_level', 'Liver_L2', 'Liver_L1', 'Kidney_L1',  'Kidney_L2', \n",
    "                  'Lung_v2_L1', 'Lung_v2_L2', 'Lung_v2_L3', 'Lung_v2_L4', 'Lung_v2_L5', 'Pancreas_L1']\n",
    "celltypist_organs = ['intestine_L1', 'kidney_L1', 'liver_L1', 'lung_L1', 'pancreas_L1', 'spleen_L1', \n",
    "                     'Adult_Human_Skin_pkl', 'Healthy_Human_Liver_pkl', 'Adult_Human_PancreaticIslet_pkl', \n",
    "                     'Human_Lung_Atlas_pkl']\n",
    "popv_organs = ['large intestine', 'liver', 'lung', 'male reproductive system', 'pancreas', 'prostate gland', \n",
    "               'respiratory system', 'skin', 'small intestine', 'spleen', 'thymus']\n",
    "\n",
    "# Filter the combined dataframe for each annotation method and their corresponding organ levels\n",
    "filtered_azimuth = final_output_combined[(final_output_combined['annotation_method'] == 'azimuth') & \n",
    "                                         (final_output_combined['Organ_Level'].isin(azimuth_organs))]\n",
    "\n",
    "filtered_celltypist = final_output_combined[(final_output_combined['annotation_method'] == 'celltypist') & \n",
    "                                            (final_output_combined['Organ_Level'].isin(celltypist_organs))]\n",
    "\n",
    "filtered_popv = final_output_combined[(final_output_combined['annotation_method'] == 'popv') & \n",
    "                                      (final_output_combined['Organ_Level'].isin(popv_organs))]\n",
    "\n",
    "# Combine the filtered dataframes\n",
    "final_filtered_combined = pd.concat([filtered_azimuth, filtered_celltypist, filtered_popv], ignore_index=True)\n",
    "\n",
    "# Save the combined dataframe to a CSV file\n",
    "output_path = 'output/filtered-CTs-with-datasets-with-organ.csv'\n",
    "final_filtered_combined.to_csv(output_path, index=False)\n",
    "\n",
    "final_filtered_combined.head()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "3c71c047",
   "metadata": {},
   "source": [
    "###  Additonal work"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 4,
   "id": "f0b04f0f",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Azimuth Organs: ['Heart_L2', 'Kidney_L3', 'Lung_v2_finest_level', 'Liver_L2', 'Liver_L1', 'Heart_L1', 'Kidney_L1', 'Human_PBMC_L3', 'Human_PBMC_L1', 'Kidney_L2', 'Lung_v2_L1', 'Lung_v2_L2', 'Lung_v2_L3', 'Lung_v2_L4', 'Lung_v2_L5', 'Pancreas_L1', 'Human_PBMC_L2', 'Bone_marrow_L1', 'Bone_marrow_L2', 'Adipose_L1', 'Adipose_L2', 'Tonsil_v2_L2', 'Tonsil_v2_L1']\n",
      "\n",
      "Celltypist Organs: ['blood_L1', 'bone marrow_L1', 'heart_L1', 'hippocampus_L1', 'intestine_L1', 'kidney_L1', 'liver_L1', 'lung_L1', 'lymph node_L1', 'pancreas_L1', 'skeletal muscle_L1', 'spleen_L1', 'Adult_Human_Skin_pkl', 'Healthy_Human_Liver_pkl', 'Adult_Human_PancreaticIslet_pkl', 'Human_Lung_Atlas_pkl', 'Healthy_Adult_Heart_pkl', 'Human_AdultAged_Hippocampus_pkl']\n",
      "\n",
      "Popv Organs: ['blood', 'blood vasculature', 'bone marrow', 'eye', 'heart', 'large intestine', 'liver', 'lung', 'lymph node', 'male reproductive system', 'mammary gland', 'mesenteric lymph node', 'pancreas', 'prostate gland', 'respiratory system', 'skin', 'small intestine', 'spleen', 'thymus', 'trachea', 'urinary bladder', 'uterus']\n"
     ]
    }
   ],
   "source": [
    "# Load the files\n",
    "azimuth = pd.read_csv(\"C:\\\\Users\\\\Supriya\\\\Downloads\\\\azimuth.csv\")\n",
    "celltypist = pd.read_csv(\"C:\\\\Users\\\\Supriya\\\\Downloads\\\\celltypist.csv\")\n",
    "popv = pd.read_csv(\"C:\\\\Users\\\\Supriya\\\\Downloads\\\\popv.csv\")\n",
    "\n",
    "# Get the unique organ levels for each file\n",
    "azimuth_organs = azimuth['Organ_Level'].unique().tolist()\n",
    "celltypist_organs = celltypist['Organ_Level'].unique().tolist()\n",
    "popv_organs = popv['Organ_Level'].unique().tolist()\n",
    "\n",
    "print(\"Azimuth Organs:\", azimuth_organs)\n",
    "print()\n",
    "print(\"Celltypist Organs:\", celltypist_organs)\n",
    "print()\n",
    "print(\"Popv Organs:\", popv_organs)"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "4ca9410e",
   "metadata": {},
   "outputs": [],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# Load the JSON file\n",
    "file_path = 'C:\\\\Users\\\\Supriya\\\\Downloads\\\\atlas-enriched-dataset-graph.jsonld'\n",
    "\n",
    "with open(file_path, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "# Provided CT Labels and CT IDs\n",
    "ct_labels_ids = {\n",
    "    \"glomerular capillary endothelial cell\": \"CL:1001005\",\n",
    "    \"efferent arteriole endothelial cell\": \"CL:1001099\",\n",
    "    \"afferent arteriole endothelial cell\": \"CL:1001096\",\n",
    "    \"peritubular capillary endothelial cell\": \"CL:1001033\",\n",
    "    \"vasa recta ascending limb cell\": \"CL:1001131\",\n",
    "    \"vasa recta descending limb cell\": \"CL:1001285\",\n",
    "    \"alveolar capillary type 1 endothelial cell\": \"CL:4028002\",\n",
    "    \"capillary endothelial cell\": \"CL:0002144\",\n",
    "    \"blood vessel smooth muscle cell\": \"CL:0019018\",\n",
    "    \"endothelial cell of artery\": \"CL:1000413\",\n",
    "    \"vein endothelial cell\": \"CL:0002543\",\n",
    "    \"endothelial cell of hepatic sinusoid\": \"CL:1000398\",\n",
    "    \"prostate gland microvascular endothelial cell\": \"CL:2000059\",\n",
    "    \"blood vessel endothelial cell\": \"CL:0000071\",\n",
    "    \"splenic endothelial cell\": \"CL:2000053\"\n",
    "}\n",
    "\n",
    "# Identifying the structure within the \"@graph\" key to process accordingly\n",
    "graph_data = json_data[\"@graph\"]"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 29,
   "id": "88dc22be",
   "metadata": {},
   "outputs": [],
   "source": [
    "# Grouping ensembl_ids by cell_id and adding cell_label\n",
    "grouped_matches = {}\n",
    "for item in graph_data:\n",
    "    samples = item.get(\"samples\", [])\n",
    "    for sample in samples:\n",
    "        sections = sample.get(\"sections\", [])\n",
    "        for section in sections:\n",
    "            datasets = section.get(\"datasets\", [])\n",
    "            for dataset in datasets:\n",
    "                summaries = dataset.get(\"summaries\", [])\n",
    "                for summary in summaries:\n",
    "                    sum_details = summary.get(\"summary\", [])\n",
    "                    for sum_detail in sum_details:\n",
    "                        cell_id = sum_detail.get(\"cell_id\")\n",
    "                        gene_expr_list = sum_detail.get(\"gene_expr\", [])\n",
    "                        if cell_id in ct_labels_ids.values():\n",
    "                            if isinstance(gene_expr_list, list):\n",
    "                                for gene_expr in gene_expr_list:\n",
    "                                    if isinstance(gene_expr, dict):\n",
    "                                        ensembl_id = gene_expr.get(\"ensembl_id\")\n",
    "                                        if cell_id not in grouped_matches:\n",
    "                                            grouped_matches[cell_id] = {\"cell_label\": \"\", \"ensembl_ids\": []}\n",
    "                                        grouped_matches[cell_id][\"cell_label\"] = [label for label, id in ct_labels_ids.items() if id == cell_id][0]\n",
    "                                        grouped_matches[cell_id][\"ensembl_ids\"].append(ensembl_id)\n",
    "\n",
    "# Preparing the final dataframe\n",
    "final_matches = []\n",
    "for cell_id, details in grouped_matches.items():\n",
    "    final_matches.append({\n",
    "        \"cell_id\": cell_id,\n",
    "        \"cell_label\": details[\"cell_label\"],\n",
    "        \"ensembl_ids\": \", \".join(details[\"ensembl_ids\"])\n",
    "    })\n",
    "\n",
    "final_matches_df = pd.DataFrame(final_matches)\n",
    "\n",
    "# Displaying the final dataframe\n",
    "final_matches_df.to_csv('Biomarker_for_Vasculature_CTs.csv')"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 28,
   "id": "f66ff2e1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['CL:1001131', 'CL:1001005', 'CL:0002144', 'CL:1000413',\n",
       "       'CL:0002543', 'CL:1000398'], dtype=object)"
      ]
     },
     "execution_count": 28,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "final_matches_df['cell_id'].unique()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 59,
   "id": "cd987cc1",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>organ_id</th>\n",
       "      <th>annotation_method</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>mean_expr_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000130300.9</td>\n",
       "      <td>5.059016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000169744.13</td>\n",
       "      <td>40.209354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000072163.20</td>\n",
       "      <td>4.431148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000150625.16</td>\n",
       "      <td>13.383242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000251322.9</td>\n",
       "      <td>3.683606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cell_id                      cell_label  \\\n",
       "0  CL:1001131  vasa recta ascending limb cell   \n",
       "1  CL:1001131  vasa recta ascending limb cell   \n",
       "2  CL:1001131  vasa recta ascending limb cell   \n",
       "3  CL:1001131  vasa recta ascending limb cell   \n",
       "4  CL:1001131  vasa recta ascending limb cell   \n",
       "\n",
       "                                        organ_id annotation_method  \\\n",
       "0  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "1  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "2  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "3  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "4  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "\n",
       "           ensembl_id  mean_expr_value  \n",
       "0   ENSG00000130300.9         5.059016  \n",
       "1  ENSG00000169744.13        40.209354  \n",
       "2  ENSG00000072163.20         4.431148  \n",
       "3  ENSG00000150625.16        13.383242  \n",
       "4   ENSG00000251322.9         3.683606  "
      ]
     },
     "execution_count": 59,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "import json\n",
    "import pandas as pd\n",
    "\n",
    "# Load the JSON file\n",
    "file_path = 'ref_data/atlas-enriched-dataset-graph.jsonld'\n",
    "\n",
    "with open(file_path, 'r') as file:\n",
    "    json_data = json.load(file)\n",
    "\n",
    "# Provided CT Labels and CT IDs\n",
    "ct_labels_ids = {\n",
    "    \"glomerular capillary endothelial cell\": \"CL:1001005\",\n",
    "    \"efferent arteriole endothelial cell\": \"CL:1001099\",\n",
    "    \"afferent arteriole endothelial cell\": \"CL:1001096\",\n",
    "    \"peritubular capillary endothelial cell\": \"CL:1001033\",\n",
    "    \"vasa recta ascending limb cell\": \"CL:1001131\",\n",
    "    \"vasa recta descending limb cell\": \"CL:1001285\",\n",
    "    \"alveolar capillary type 1 endothelial cell\": \"CL:4028002\",\n",
    "    \"capillary endothelial cell\": \"CL:0002144\",\n",
    "    \"blood vessel smooth muscle cell\": \"CL:0019018\",\n",
    "    \"endothelial cell of artery\": \"CL:1000413\",\n",
    "    \"vein endothelial cell\": \"CL:0002543\",\n",
    "    \"endothelial cell of hepatic sinusoid\": \"CL:1000398\",\n",
    "    \"prostate gland microvascular endothelial cell\": \"CL:2000059\",\n",
    "    \"blood vessel endothelial cell\": \"CL:0000071\",\n",
    "    \"splenic endothelial cell\": \"CL:2000053\"\n",
    "}\n",
    "\n",
    "# Identifying the structure within the \"@graph\" key to process accordingly\n",
    "graph_data = json_data[\"@graph\"]\n",
    "\n",
    "# Grouping ensembl_ids by cell_id and organ_id, adding cell_label and mean_gene_expr_value\n",
    "grouped_matches = {}\n",
    "for item in graph_data:\n",
    "    samples = item.get(\"samples\", [])\n",
    "    for sample in samples:\n",
    "        sections = sample.get(\"sections\", [])\n",
    "        for section in sections:\n",
    "            datasets = section.get(\"datasets\", [])\n",
    "            for dataset in datasets:\n",
    "                summaries = dataset.get(\"summaries\", [])               \n",
    "                for summary in summaries:\n",
    "                    annotation_method = summary.get(\"annotation_method\", \"Unknown\")\n",
    "                    sum_details = summary.get(\"summary\", [])\n",
    "                    for sum_detail in sum_details:\n",
    "                        cell_id = sum_detail.get(\"cell_id\")\n",
    "                        gene_expr_list = sum_detail.get(\"gene_expr\", [])\n",
    "                        organ_id = dataset.get(\"organ_id\", \"Unknown\")\n",
    "                        if cell_id in ct_labels_ids.values():\n",
    "                            if isinstance(gene_expr_list, list) and gene_expr_list:\n",
    "                                for gene_expr in gene_expr_list:\n",
    "                                    if isinstance(gene_expr, dict):\n",
    "                                        ensembl_id = gene_expr.get(\"ensembl_id\")\n",
    "                                        mean_expr_value = gene_expr.get(\"mean_gene_expr_value\")\n",
    "                                        key = (cell_id, organ_id, annotation_method)\n",
    "                                        if key not in grouped_matches:\n",
    "                                            grouped_matches[key] = {\"cell_label\": \"\", \"ensembl_ids\": [], \"mean_expr_values\": []}\n",
    "                                        grouped_matches[key][\"cell_label\"] = [label for label, id in ct_labels_ids.items() if id == cell_id][0]\n",
    "                                        grouped_matches[key][\"ensembl_ids\"].append(ensembl_id)\n",
    "                                        grouped_matches[key][\"mean_expr_values\"].append(mean_expr_value)\n",
    "                            else:\n",
    "                                # Handle entries with empty gene_expr lists\n",
    "                                key = (cell_id, organ_id, annotation_method)\n",
    "                                if key not in grouped_matches:\n",
    "                                    grouped_matches[key] = {\"cell_label\": \"\", \"ensembl_ids\": [], \"mean_expr_values\": []}\n",
    "                                grouped_matches[key][\"cell_label\"] = [label for label, id in ct_labels_ids.items() if id == cell_id][0]\n",
    "\n",
    "# Preparing the final dataframe\n",
    "final_matches = []\n",
    "for (cell_id, organ_id, annotation_method), details in grouped_matches.items():\n",
    "    final_matches.append({\n",
    "        \"cell_id\": cell_id,\n",
    "        \"cell_label\": details[\"cell_label\"],\n",
    "        \"organ_id\": organ_id,\n",
    "        \"annotation_method\": annotation_method,\n",
    "        \"ensembl_ids\": \", \".join(details[\"ensembl_ids\"]),\n",
    "        \"mean_expr_values\": \", \".join(map(str, details[\"mean_expr_values\"]))\n",
    "    })\n",
    "\n",
    "final_matches_df = pd.DataFrame(final_matches)\n",
    "\n",
    "# Expanding the dataframe to have each row for each ensembl_id\n",
    "expanded_matches = []\n",
    "for _, row in final_matches_df.iterrows():\n",
    "    cell_id = row[\"cell_id\"]\n",
    "    cell_label = row[\"cell_label\"]\n",
    "    organ_id = row[\"organ_id\"]\n",
    "    annotation_method = row[\"annotation_method\"]\n",
    "    ensembl_ids = row[\"ensembl_ids\"].split(\", \")\n",
    "    mean_expr_values = row[\"mean_expr_values\"].split(\", \")\n",
    "    \n",
    "    for ensembl_id, mean_expr_value in zip(ensembl_ids, mean_expr_values):\n",
    "        try:\n",
    "            mean_expr_value_float = float(mean_expr_value)\n",
    "        except ValueError:\n",
    "            mean_expr_value_float = None  # Handle non-numeric values\n",
    "        expanded_matches.append({\n",
    "            \"cell_id\": cell_id,\n",
    "            \"cell_label\": cell_label,\n",
    "            \"organ_id\": organ_id,\n",
    "            \"annotation_method\": annotation_method,\n",
    "            \"ensembl_id\": ensembl_id,\n",
    "            \"mean_expr_value\": mean_expr_value_float\n",
    "        })\n",
    "\n",
    "expanded_matches_df = pd.DataFrame(expanded_matches)\n",
    "\n",
    "# Displaying the expanded dataframe\n",
    "expanded_matches_df.head()  # Displaying only the first few rows"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 61,
   "id": "4f122e1b",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['http://purl.obolibrary.org/obo/UBERON_0002113',\n",
       " 'http://purl.obolibrary.org/obo/UBERON_0000948',\n",
       " 'http://purl.obolibrary.org/obo/UBERON_0001255',\n",
       " 'http://purl.obolibrary.org/obo/UBERON_0002048',\n",
       " 'http://purl.obolibrary.org/obo/UBERON_0002107']"
      ]
     },
     "execution_count": 61,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Extract unique organ IDs from the expanded dataframe\n",
    "unique_organ_ids = expanded_matches_df[\"organ_id\"].unique()\n",
    "\n",
    "unique_organ_ids_list = unique_organ_ids.tolist()\n",
    "unique_organ_ids_list"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 62,
   "id": "42f8cbe0",
   "metadata": {},
   "outputs": [],
   "source": [
    "organ_id_to_organ_name = {\n",
    "         'http://purl.obolibrary.org/obo/UBERON_0002113' : 'Kidney',\n",
    "         'http://purl.obolibrary.org/obo/UBERON_0000948' : 'heart',\n",
    "         'http://purl.obolibrary.org/obo/UBERON_0001255': 'urinary bladder',\n",
    "         'http://purl.obolibrary.org/obo/UBERON_0002048' : 'lung',\n",
    "         'http://purl.obolibrary.org/obo/UBERON_0002107' : 'liver'\n",
    "        }"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 63,
   "id": "fac496df",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>cell_id</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>organ_id</th>\n",
       "      <th>annotation_method</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>mean_expr_value</th>\n",
       "      <th>organ_name</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000130300.9</td>\n",
       "      <td>5.059016</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000169744.13</td>\n",
       "      <td>40.209354</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000072163.20</td>\n",
       "      <td>4.431148</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000150625.16</td>\n",
       "      <td>13.383242</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>ENSG00000251322.9</td>\n",
       "      <td>3.683606</td>\n",
       "      <td>Kidney</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "      cell_id                      cell_label  \\\n",
       "0  CL:1001131  vasa recta ascending limb cell   \n",
       "1  CL:1001131  vasa recta ascending limb cell   \n",
       "2  CL:1001131  vasa recta ascending limb cell   \n",
       "3  CL:1001131  vasa recta ascending limb cell   \n",
       "4  CL:1001131  vasa recta ascending limb cell   \n",
       "\n",
       "                                        organ_id annotation_method  \\\n",
       "0  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "1  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "2  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "3  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "4  http://purl.obolibrary.org/obo/UBERON_0002113           azimuth   \n",
       "\n",
       "           ensembl_id  mean_expr_value organ_name  \n",
       "0   ENSG00000130300.9         5.059016     Kidney  \n",
       "1  ENSG00000169744.13        40.209354     Kidney  \n",
       "2  ENSG00000072163.20         4.431148     Kidney  \n",
       "3  ENSG00000150625.16        13.383242     Kidney  \n",
       "4   ENSG00000251322.9         3.683606     Kidney  "
      ]
     },
     "execution_count": 63,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Adding a new column to the dataframe\n",
    "expanded_matches_df['organ_name'] = expanded_matches_df['organ_id'].map(organ_id_to_organ_name)\n",
    "\n",
    "# Display the first few rows of the updated dataframe\n",
    "expanded_matches_df.head() "
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 64,
   "id": "0e47f431",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/html": [
       "<div>\n",
       "<style scoped>\n",
       "    .dataframe tbody tr th:only-of-type {\n",
       "        vertical-align: middle;\n",
       "    }\n",
       "\n",
       "    .dataframe tbody tr th {\n",
       "        vertical-align: top;\n",
       "    }\n",
       "\n",
       "    .dataframe thead th {\n",
       "        text-align: right;\n",
       "    }\n",
       "</style>\n",
       "<table border=\"1\" class=\"dataframe\">\n",
       "  <thead>\n",
       "    <tr style=\"text-align: right;\">\n",
       "      <th></th>\n",
       "      <th>organ_name</th>\n",
       "      <th>organ_name_in_FTU</th>\n",
       "      <th>organ_id</th>\n",
       "      <th>annotation_method</th>\n",
       "      <th>FTU_name</th>\n",
       "      <th>cell_id</th>\n",
       "      <th>cell_label</th>\n",
       "      <th>ensembl_id</th>\n",
       "      <th>mean_expr_value</th>\n",
       "    </tr>\n",
       "  </thead>\n",
       "  <tbody>\n",
       "    <tr>\n",
       "      <th>0</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>Thick Ascending Loop Of Henle</td>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>ENSG00000130300.9</td>\n",
       "      <td>5.059016</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>1</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>Thick Ascending Loop Of Henle</td>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>ENSG00000169744.13</td>\n",
       "      <td>40.209354</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>2</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>Thick Ascending Loop Of Henle</td>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>ENSG00000072163.20</td>\n",
       "      <td>4.431148</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>3</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>Thick Ascending Loop Of Henle</td>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>ENSG00000150625.16</td>\n",
       "      <td>13.383242</td>\n",
       "    </tr>\n",
       "    <tr>\n",
       "      <th>4</th>\n",
       "      <td>Kidney</td>\n",
       "      <td>Kidney</td>\n",
       "      <td>http://purl.obolibrary.org/obo/UBERON_0002113</td>\n",
       "      <td>azimuth</td>\n",
       "      <td>Thick Ascending Loop Of Henle</td>\n",
       "      <td>CL:1001131</td>\n",
       "      <td>vasa recta ascending limb cell</td>\n",
       "      <td>ENSG00000251322.9</td>\n",
       "      <td>3.683606</td>\n",
       "    </tr>\n",
       "  </tbody>\n",
       "</table>\n",
       "</div>"
      ],
      "text/plain": [
       "  organ_name organ_name_in_FTU                                       organ_id  \\\n",
       "0     Kidney            Kidney  http://purl.obolibrary.org/obo/UBERON_0002113   \n",
       "1     Kidney            Kidney  http://purl.obolibrary.org/obo/UBERON_0002113   \n",
       "2     Kidney            Kidney  http://purl.obolibrary.org/obo/UBERON_0002113   \n",
       "3     Kidney            Kidney  http://purl.obolibrary.org/obo/UBERON_0002113   \n",
       "4     Kidney            Kidney  http://purl.obolibrary.org/obo/UBERON_0002113   \n",
       "\n",
       "  annotation_method                       FTU_name     cell_id  \\\n",
       "0           azimuth  Thick Ascending Loop Of Henle  CL:1001131   \n",
       "1           azimuth  Thick Ascending Loop Of Henle  CL:1001131   \n",
       "2           azimuth  Thick Ascending Loop Of Henle  CL:1001131   \n",
       "3           azimuth  Thick Ascending Loop Of Henle  CL:1001131   \n",
       "4           azimuth  Thick Ascending Loop Of Henle  CL:1001131   \n",
       "\n",
       "                       cell_label          ensembl_id  mean_expr_value  \n",
       "0  vasa recta ascending limb cell   ENSG00000130300.9         5.059016  \n",
       "1  vasa recta ascending limb cell  ENSG00000169744.13        40.209354  \n",
       "2  vasa recta ascending limb cell  ENSG00000072163.20         4.431148  \n",
       "3  vasa recta ascending limb cell  ENSG00000150625.16        13.383242  \n",
       "4  vasa recta ascending limb cell   ENSG00000251322.9         3.683606  "
      ]
     },
     "execution_count": 64,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "# Load the new CSV file\n",
    "ftu_file_path = 'ref_data/FTU Cell Count Table - Cell_Type_Count.csv'\n",
    "ftu_df = pd.read_csv(ftu_file_path)\n",
    "\n",
    "# Select and clean up relevant columns from FTU dataframe\n",
    "ftu_cleaned_df = ftu_df[['Organ', 'CT Label in CL', 'FTU Label in Uberon']].dropna().drop_duplicates()\n",
    "\n",
    "# Ensure unique mapping by dropping duplicates based on 'CT Label in CL'\n",
    "ftu_unique_df = ftu_cleaned_df.drop_duplicates(subset=['CT Label in CL'])\n",
    "\n",
    "# Perform the match and add the new column for FTU name\n",
    "expanded_matches_df['FTU_name'] = expanded_matches_df['cell_label'].map(\n",
    "    ftu_unique_df.set_index('CT Label in CL')['FTU Label in Uberon']\n",
    ")\n",
    "\n",
    "# Adding the organ name from FTU data\n",
    "expanded_matches_df['Organ_name_FTU'] = expanded_matches_df['cell_label'].map(\n",
    "    ftu_unique_df.set_index('CT Label in CL')['Organ']\n",
    ")\n",
    "\n",
    "# Rearrange columns as per the requirement\n",
    "expanded_matches_df = expanded_matches_df[[\n",
    "    'organ_name', 'Organ_name_FTU','organ_id', 'annotation_method', 'FTU_name', 'cell_id', 'cell_label', 'ensembl_id', 'mean_expr_value'\n",
    "]]\n",
    "\n",
    "# Rename columns for consistency\n",
    "expanded_matches_df.rename(columns={\n",
    "    'organ_name': 'organ_name',\n",
    "    'Organ_name_FTU': 'organ_name_in_FTU'\n",
    "}, inplace=True)\n",
    "\n",
    "# Display the first few rows of the updated dataframe\n",
    "expanded_matches_df.head()"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 65,
   "id": "49b6a286",
   "metadata": {},
   "outputs": [],
   "source": [
    "expanded_matches_df.to_csv('output/vasculature/gene_information-for-vasculature-CTs_1.csv', index=False)"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "e63d4534",
   "metadata": {},
   "source": [
    "#### Generate box plot images for unique cell label"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 69,
   "id": "991025b3",
   "metadata": {},
   "outputs": [],
   "source": [
    "import matplotlib.pyplot as plt\n",
    "import seaborn as sns\n",
    "\n",
    "# Create individual box plots for each cell_label\n",
    "unique_cell_labels = expanded_matches_df['cell_label'].unique()\n",
    "\n",
    "for cell_label in unique_cell_labels:\n",
    "    cell_label_data = expanded_matches_df[expanded_matches_df['cell_label'] == cell_label]\n",
    "    \n",
    "    if len(cell_label_data) > 1:  # Ensure there is enough data to plot\n",
    "        plt.figure(figsize=(8, 6))\n",
    "        sns.boxplot(data=cell_label_data, y='mean_expr_value', color='skyblue')\n",
    "        \n",
    "        # Customize the box plot\n",
    "        plt.xticks([])\n",
    "        plt.title(f'Box Plot of Mean Expression Values for {cell_label}')\n",
    "        plt.xlabel('Cell Label')\n",
    "        plt.ylabel('Mean Expression Value')\n",
    "        \n",
    "        # Adding additional visual elements\n",
    "        plt.axhline(y=cell_label_data['mean_expr_value'].median(), color='red', linestyle='-', label='Median')\n",
    "        plt.legend()\n",
    "\n",
    "        # Save each plot as an image file\n",
    "        individual_box_plot_file_path = f'output/vasculature/figures/box_plot_mean_expr_value_{cell_label.replace(\" \", \"_\")}.png'\n",
    "        plt.tight_layout()\n",
    "        plt.savefig(individual_box_plot_file_path)\n",
    "        plt.close()"
   ]
  },
  {
   "cell_type": "markdown",
   "id": "325ef524",
   "metadata": {},
   "source": [
    "#### Generate box plot images for unique cell label per annotation method"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 67,
   "id": "4a5e9368",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "array(['azimuth', 'celltypist', 'popv'], dtype=object)"
      ]
     },
     "execution_count": 67,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_annotation_methods = expanded_matches_df['annotation_method'].explode().unique()\n",
    "unique_annotation_methods"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 70,
   "id": "ba53ba84",
   "metadata": {},
   "outputs": [
    {
     "data": {
      "text/plain": [
       "['output/vasculature/figures/box_plot_mean_expr_value_vasa_recta_ascending_limb_cell_azimuth.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_glomerular_capillary_endothelial_cell_azimuth.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_capillary_endothelial_cell_azimuth.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_capillary_endothelial_cell_celltypist.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_capillary_endothelial_cell_popv.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_endothelial_cell_of_artery_azimuth.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_endothelial_cell_of_artery_celltypist.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_endothelial_cell_of_artery_popv.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_vein_endothelial_cell_celltypist.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_vein_endothelial_cell_popv.png',\n",
       " 'output/vasculature/figures/box_plot_mean_expr_value_endothelial_cell_of_hepatic_sinusoid_popv.png']"
      ]
     },
     "execution_count": 70,
     "metadata": {},
     "output_type": "execute_result"
    }
   ],
   "source": [
    "unique_annotation_methods = expanded_matches_df['annotation_method'].explode().unique()\n",
    "\n",
    "for cell_label in unique_cell_labels:\n",
    "    for annotation_method in unique_annotation_methods:\n",
    "        cell_label_data = expanded_matches_df[\n",
    "            (expanded_matches_df['cell_label'] == cell_label) &\n",
    "            (expanded_matches_df['annotation_method'] == annotation_method)\n",
    "        ]\n",
    "        \n",
    "        if len(cell_label_data) > 1:  # Ensure there is enough data to plot\n",
    "            plt.figure(figsize=(8, 6))\n",
    "            sns.boxplot(data=cell_label_data, y='mean_expr_value', color='skyblue')\n",
    "            \n",
    "            # Customize the box plot\n",
    "            plt.xticks([])\n",
    "            plt.title(f'Box Plot of Mean Expression Values for {cell_label} ({annotation_method})')\n",
    "            plt.xlabel('Cell Label')\n",
    "            plt.ylabel('Mean Expression Value')\n",
    "            \n",
    "            # Adding additional visual elements\n",
    "            plt.axhline(y=cell_label_data['mean_expr_value'].median(), color='red', linestyle='-', label='Median')\n",
    "            plt.legend()\n",
    "\n",
    "            # Save each plot as an image file\n",
    "            individual_box_plot_file_path = f'output/vasculature/figures/box_plot_mean_expr_value_{cell_label.replace(\" \", \"_\")}_{annotation_method.replace(\" \", \"_\")}.png'\n",
    "            plt.tight_layout()\n",
    "            plt.savefig(individual_box_plot_file_path)\n",
    "            plt.close()\n",
    "\n",
    "# List of generated plot file paths\n",
    "plot_file_paths = [f'output/vasculature/figures/box_plot_mean_expr_value_{cell_label.replace(\" \", \"_\")}_{annotation_method.replace(\" \", \"_\")}.png'\n",
    "                   for cell_label in unique_cell_labels\n",
    "                   for annotation_method in unique_annotation_methods\n",
    "                   if len(expanded_matches_df[(expanded_matches_df['cell_label'] == cell_label) & (expanded_matches_df['annotation_method'] == annotation_method)]) > 1]\n",
    "\n",
    "plot_file_paths"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 26,
   "id": "06aa4c46",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell IDs from the file: 12650\n"
     ]
    }
   ],
   "source": [
    "import json\n",
    "\n",
    "file_path = 'ref_data/atlas-enriched-dataset-graph.jsonld'\n",
    "\n",
    "with open(file_path, 'r') as file:\n",
    "    data = json.load(file)\n",
    "\n",
    "# Extract cell IDs from the summary section\n",
    "def extract_cell_ids(data):\n",
    "    cell_ids = []\n",
    "    for item in data['@graph']:\n",
    "        samples = item.get(\"samples\", [])\n",
    "        for sample in samples:\n",
    "            sections = sample.get(\"sections\", [])\n",
    "            for section in sections:\n",
    "                datasets = section.get(\"datasets\", [])\n",
    "                for dataset in datasets:\n",
    "                    summaries = dataset.get(\"summaries\", [])               \n",
    "                    for summary in summaries:\n",
    "                        sum_details = summary.get(\"summary\", [])\n",
    "                        for sum_detail in sum_details:\n",
    "                            cell_id = sum_detail.get(\"cell_id\")\n",
    "                            cell_ids.append(cell_id)\n",
    "    return cell_ids\n",
    "\n",
    "cell_ids = extract_cell_ids(data)\n",
    "print(f\"Cell IDs from the file: {len(cell_ids)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": 45,
   "id": "f68d0427",
   "metadata": {},
   "outputs": [
    {
     "name": "stdout",
     "output_type": "stream",
     "text": [
      "Cell IDs from the file: 6072\n"
     ]
    }
   ],
   "source": [
    "# Load the data\n",
    "file_path = 'ref_data/atlas-as-cell-summaries.jsonld'\n",
    "\n",
    "with open(file_path, 'r') as file:\n",
    "    data = json.load(file)\n",
    "\n",
    "def extract_cell_ids_summaries(data):\n",
    "    cell_ids = []\n",
    "    for item in data['@graph']:\n",
    "        summaries = item.get('summary', [])\n",
    "        for summary in summaries:\n",
    "            for detail in summary:\n",
    "                if 'cell_id' in detail:\n",
    "                    cell_ids.append(summary['cell_id'])\n",
    "    return cell_ids\n",
    "\n",
    "cell_ids_summaries = extract_cell_ids_summaries(data)\n",
    "print(f\"Cell IDs from the file: {len(cell_ids_summaries)}\")"
   ]
  },
  {
   "cell_type": "code",
   "execution_count": null,
   "id": "a5c361dc",
   "metadata": {},
   "outputs": [],
   "source": []
  }
 ],
 "metadata": {
  "kernelspec": {
   "display_name": "Python 3 (ipykernel)",
   "language": "python",
   "name": "python3"
  },
  "language_info": {
   "codemirror_mode": {
    "name": "ipython",
    "version": 3
   },
   "file_extension": ".py",
   "mimetype": "text/x-python",
   "name": "python",
   "nbconvert_exporter": "python",
   "pygments_lexer": "ipython3",
   "version": "3.11.5"
  }
 },
 "nbformat": 4,
 "nbformat_minor": 5
}
